protocols/bftcosi/bftcosi.go

```diff
@@ -0,0 +1,615 @@
+	// create response
+	resp, err := bft.prepare.CreateResponse()
+	if err != nil {
+		return err
+	}
+	// wait the verification
+	r, ok := bft.waitResponseVerification()
+	if ok {
+		// append response only if OK
+		r.Response = resp
+	}
+	dbg.Lvl3(bft.Name(), "BFTCoSi Start Response PREPARE with response:", r)
+	// send to parent
+	return bft.SendTo(bft.Parent(), r)
+}
+
+// startCommitResponse will create the response for the commit phase and send it
+// up. It will not create the response if it decided the signature is wrong from
+// the prepare phase.
+func (bft *ProtocolBFTCoSi) startResponseCommit() error {
+	r := &Response{
+		TYPE: RoundCommit,
+	}
+	// if i dont want to sign
+	if bft.signRefusal {
+		r.Exceptions = append(r.Exceptions, cosi.Exception{
+			Public:     bft.Public(),
+			Commitment: bft.commit.GetCommitment(),
+		})
+	} else {
+		// otherwise i create the response
+		resp, err := bft.commit.CreateResponse()
+		if err != nil {
+			return err
+		}
+		r.Response = resp
+	}
+	dbg.Lvl3(bft.Name(), "BFTCoSi Start Response COMMIT")
+	// send to parent
+	err := bft.SendTo(bft.Parent(), r)
+	bft.Done()
+	return err
+}
+
+// handleResponseCommit handles the responses for the commit round during the
+// response phase.
+func (bft *ProtocolBFTCoSi) handleResponseCommit(r *Response) error {
+	// check if we have enough
+	// FIXME possible data race
+	bft.tcrMut.Lock()
+	defer bft.tcrMut.Unlock()
+	bft.tempCommitResponse = append(bft.tempCommitResponse, r.Response)
+
+	if len(bft.tempCommitResponse) < len(bft.Children()) {
+		return nil
+	}
+
+	if bft.signRefusal {
+		r.Exceptions = append(r.Exceptions, cosi.Exception{
+			Public:     bft.Public(),
+			Commitment: bft.commit.GetCommitment(),
+		})
+	} else {
+		resp, err := bft.commit.Response(bft.tempCommitResponse)
+		if err != nil {
+			return err
+		}
+		r.Response = resp
+	}
+
+	// notify we have finished to participate in this signature
+	bft.doneSigning <- true
+	dbg.Lvl3(bft.Name(), "BFTCoSi handle Response COMMIT (refusal=", bft.signRefusal, ")")
+	// if root we have finished
+	if bft.IsRoot() {
+		sig := bft.Signature()
+		if bft.onSignatureDone != nil {
+			bft.onSignatureDone(sig)
+		}
+		bft.Done()
+		return nil
+	}
+
+	// otherwise , send the response up
+	err := bft.SendTo(bft.Parent(), r)
+	bft.Done()
+	return err
+}
+
+func (bft *ProtocolBFTCoSi) handleResponsePrepare(r *Response) error {
+	// check if we have enough
+	bft.tprMut.Lock()
+	defer bft.tprMut.Unlock()
+	bft.tempPrepareResponse = append(bft.tempPrepareResponse, r.Response)
+	if len(bft.tempPrepareResponse) < len(bft.Children()) {
+		return nil
+	}
+
+	// wait for verification
+	bzrReturn, ok := bft.waitResponseVerification()
+	if ok {
+		// append response
+		resp, err := bft.prepare.Response(bft.tempPrepareResponse)
+		if err != nil {
+			return err
+		}
+		bzrReturn.Response = resp
+	}
+
+	dbg.Lvl3("BFTCoSi Handle Response PREPARE")
+	// if I'm root, we are finished, let's notify the "commit" round
```

Comment:
Comments about the true if-branch should be inside the if-branch. Here it's 50/50 ;) Either outside

`// Check whether we need to notify the commit round`

or inside

`// Notify 'commit'-round as we're root`